### PR TITLE
feat(index): Provide a similar API to Query

### PR DESCRIPTION
**Pull request by adrien2p**
**What**
Align the index engine API to be similar to the Query API

## Example

```ts
        // Benefit from the same level of typing like the remote query

        const { data, metadata } = await indexEngine.query<'product'>({
          fields: [
            "product.*",
            "product.variants.*",
            "product.variants.prices.*",
          ],
          filters: {
            product: {
              variants: {
                prices: {
                  amount: { $gt: 50 },
                },
              },
            },
          },
          pagination: {
            order: {
              product: {
                variants: {
                  prices: {
                    amount: "DESC",
                  },
                },
              },
            },
          },
        })
```

### CODE REVIEW

File: packages/modules/index/src/utils/flatten-object-keys.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * @param where
+ */
+export function flattenObjectKeys(where: Record<string, any>) {
+  const result: Record<string, any> = {}
+
+  function flatten(obj: Record<string, any>, path?: string) {
+    for (const key in obj) {
+      const isOperatorMap = !!OPERATOR_MAP[key]
+
+      if (!isOperatorMap) {
```
Comment (carlos-r-l-rodrigues) [inline packages/modules/index/src/utils/flatten-object-keys.ts]:
```suggestion
      if (isOperatorMap) {
        result[path ?? key] = obj
        continur
      }
      
      // ...
```

File: packages/core/types/src/index/query-config/query-input-config.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    ? string[]
+    : ObjectToIndexFields<
+        IndexServiceEntryPoints[TEntry & keyof IndexServiceEntryPoints]
+      >[]
+  filters?: IndexFilters<TEntry>
+  joinFilters?: IndexFilters<TEntry>
+  pagination?: {
+    skip?: number
+    take?: number
+    orderBy?: IndexOrderBy<TEntry>
```
Comment (olivermrbl) [inline packages/core/types/src/index/query-config/query-input-config.ts]: question: Wasn't this just `order` in `remoteQuery`?

Comment (adrien2p) [reply packages/core/types/src/index/query-config/query-input-config.ts]: my bad, let me chnge that

File: packages/core/types/src/index/service.ts [hunk 3]
```diff
@@ -1,6 +1,11 @@
 import { IModuleService } from "../modules-sdk"
+import { IndexQueryConfig } from "./query-config"
 
 export interface IIndexService extends IModuleService {
-  query(...args): Promise<any>
-  queryAndCount(...args): Promise<any>
+  query<const TEntry extends string>(
+    config: IndexQueryConfig<TEntry>
+  ): Promise<any[]>
+  queryAndCount<const TEntry extends string, Test = IndexQueryConfig<TEntry>>(
```
Comment (olivermrbl) [inline packages/core/types/src/index/service.ts]: question: Can we kill this `queryAndCount` and return the same shape as `query.graph`?

```
const { data, metadata } = ...
``` 

Count would be part of metadata. Wdyt?

Comment (adrien2p) [reply packages/core/types/src/index/service.ts]: Yes we can def do that 👍 i dont see any issues

File: packages/modules/index/package.json [hunk 4]
```diff
@@ -23,9 +23,9 @@
   "scripts": {
     "watch": "tsc --build --watch",
     "watch:test": "tsc --build tsconfig.spec.json --watch",
-    "build": "rimraf dist && tsc --noEmit && tsc -p ./tsconfig.build.json && tsc-alias -p ./tsconfig.build.json",
+    "build": "rimraf dist && tsc -p ./tsconfig.spec.json --noEmit && tsc -p ./tsconfig.build.json && tsc-alias -p ./tsconfig.build.json",
     "test": "jest --passWithNoTests ./src",
-    "test:integration": "jest --runInBand --forceExit -- integration-tests/**/__tests__/**/*.ts",
+    "test:integration": "jest --runInBand --forceExit -- integration-tests/__tests__/index-engine-module.spec.ts",
```
Comment (olivermrbl) [inline packages/modules/index/package.json]: todo: rm

File: packages/modules/index/src/services/index-module-service.ts [hunk 5]
```diff
@@ -93,20 +94,23 @@ export default class IndexModuleService implements IndexTypes.IIndexService {
...
+  ) {
+    return await this.storageProvider_.query(config)
   }
 
-  async queryAndCount(...args) {
-    return await this.storageProvider_.queryAndCount.apply(
-      this.storageProvider_,
-      args
-    )
+  async queryAndCount<const TEntry extends string>(
```
Comment (olivermrbl) [inline packages/modules/index/src/services/index-module-service.ts]: todo(if-possible): Remove in favor of unified return shape as discussed above

File: packages/modules/index/src/types/index.ts [hunk 6]
```diff
@@ -119,9 +120,13 @@ export interface StorageProvider {
 
   onApplicationStart?(): Promise<void>
 
-  query(...args): unknown
+  query<const TEntry extends string>(
+    config: IndexTypes.IndexQueryConfig<TEntry>
+  ): Promise<any[]>
 
-  queryAndCount(...args): unknown
+  queryAndCount<const TEntry extends string>(
```
Comment (olivermrbl) [inline packages/modules/index/src/types/index.ts]: same

**Review by olivermrbl — APPROVED** [review]
Nice work, few comments 💪

File: packages/modules/index/src/services/postgres-provider.ts [hunk 7]
```diff
@@ -353,6 +227,91 @@ export class PostgresProvider {
...
+    const { take, skip, order: inputOrderBy = {} } = config.pagination ?? {}
+
+    const select = normalizeFieldsSelection(fields)
+    const where = flattenObjectKeys(filters)
+    const joinWhere = flattenObjectKeys(joinFilters)
+    const orderBy = flattenObjectKeys(inputOrderBy)
+
+    const { manager } = sharedContext as { manager: SqlEntityManager }
+    let hasPagination = false
+    if (typeof take === "number" || typeof skip === "number") {
```
Comment (carlos-r-l-rodrigues) [inline packages/modules/index/src/services/postgres-provider.ts]: I believe we can have a `take` that is not a pagination.
I would only consider to paginate if we have `skip`.

**Review by carlos-r-l-rodrigues — APPROVED** [review]
🔥 🔥 🔥

**Pull request merged**